### PR TITLE
Make FindPackage(DD4hep) work even if thisdd4hep.sh was not sourced

### PR DESCRIPTION
**Pull request by HadrienG2**
A long time ago, one could use DD4hep from CMake without needing to source the `thisdd4hep.sh` script, provided that the required environment was set in another way (e.g. by installing to `/usr/local`)

Then commit https://github.com/AIDASoft/DD4hep/commit/8e5b6b689f69bedd1c90532d9d86995d93462755 came and introduced a way to make DD4hep more relocatable, at the cost of making use of `thisdd4hep.sh` a hard requirement for CMake's `FindPackage(DD4hep)` to work at all.

I think we can have our cake and eat it too here. The trick is to check if the environment variables that are configured by `thisdd4hep.sh` have been set. If they were, then we can use them as before to make DD4hep more relocatable. If they weren't set, we can fallback to the old non-relocatable workflow.

BEGINRELEASENOTES

- Make FindPackage(DD4hep) work even if thisdd4hep.sh was not sourced

ENDRELEASENOTES

### CODE REVIEW

File: cmake/DD4hepConfig.cmake.in [hunk 1]
```diff
@@ -13,8 +13,16 @@
...
-set ( DD4hep_ROOT        "$ENV{DD4hep_ROOT}" )
+if( DEFINED ENV{DD4hep_DIR} )
+  set ( DD4hep_DIR         "$ENV{DD4hep_DIR}" )
+else()
+  set ( DD4hep_DIR         "@CMAKE_INSTALL_PREFIX@" )
+endif()
+if( DEFINED ENV{DD4hep_ROOT} )
+  set ( DD4hep_ROOT        "$ENV{DD4hep_ROOT}" )
+else()
+  set ( DD4hep_ROOT         ${DD4hep_DIR} )
```
Comment (petricm) [inline cmake/DD4hepConfig.cmake.in]: I think you want `@CMAKE_INSTALL_PREFIX@` not `${DD4hep_DIR}`

Comment (HadrienG2) [reply cmake/DD4hepConfig.cmake.in]: You are right, otherwise it is not robust if the user overrides DD4hep_DIR but not DD4hep_ROOT. Fixed.

**Comment by MarkusFrankATcernch** [comment]
Is everybody convinced that we can have the cake and also eat it?
This is an important change here and it should not break what is present.

**Comment by HadrienG2** [comment]
Here is why I think this works and should not break existing use cases:

- If the DD4hep_DIR and DD4hep_ROOT environment variables are defined, then the behaviour is the same as in the current master (you can check this by comparing the "if DEFINED" code paths with the original code).
- If one of these environment variables are not defined, then the code on master does not work (it tries to access a nonexistent filesystem path and crashes). This PR makes it work in the case where the DD4hep install has not been relocated.

**Comment by petricm** [comment]
We had a discussion and we think the best solution would be to solve this by determining what is the location of the `DD4hepConfig.cmake` file and setting based on that the two variables in question e.g.

```cmake
get_filename_component(_thisdir "${CMAKE_CURRENT_LIST_FILE}" PATH)
```

but you have to take into account that where are two `DD4hepConfig.cmake` files installed: one in base and one in cmake.

**Comment by petricm** [comment]
Please get rid of the `ENV{DD4hep_DIR}` and `ENV{DD4hep_ROOT}` in the config file, they are not really needed in this way.

**Comment by HadrienG2** [comment]
Done :)

**Pull request merged**